### ISSUE TITLE
Johnson shortest path functions should have a mode argument

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#2193]
**What is the feature or improvement you would like to see?**

`igraph_distances_johnson()` and `igraph_shortest_paths_johnson()` should have a `mode` argument.

**Use cases for the feature**

The Dijkstra, Bellman-Ford and unweighted variants already have this argument. Ideally, a high-level interface would have a single distances() function, where the algorithm and the mode would be separate parameters. Right now, the Johnson algorithm cannot handle the mode.

**Comment by GroteGnoom** [comment]
One problem here is that this function is only useful when there are negative weights for edges, and IGRAPH_ALL will then always cause a negative loop because you can go back and forth over the same edge.

So even with one negative non-loop edge you get this:
```
Error at src/paths/bellman_ford.c:159 : Negative loop in graph while calculating distances with Bellman-Ford algorithm. - Negative loop detected while calculating shortest paths.
``` 

That's also why we have this:
```
    if (!igraph_is_directed(graph)) {
        IGRAPH_ERROR("Johnson's shortest path: undirected graph and negative weight.",
                     IGRAPH_EINVAL);
    }
```

I will add a check for mode == IGRAPH_ALL to this.

**Comment by szhorvat** [comment]
Since this is a breaking change, for 0.11 only, it might be a good idea to wait a bit with fixing this. The fix should go on `develop`, not `master`.

**Comment by GroteGnoom** [comment]
Fixed in #2218
